File: public/python/089-reorganize-string.py
# LeetCode Problem #767: Reorganize String
# Difficulty: Medium
# Link: https://leetcode.com/problems/reorganize-string/

import heapq

# ─────────────────────────────────────────────
# APPROACH 1: Max Heap Greedy (Pair Pop) | O(n log 26) time | O(1) space
# EXPLAIN: Each step pop the two most frequent characters, append both, and re-push with decremented counts.
# WHEN: Clean greedy approach; handles the impossible case naturally when one char remains with freq > 1.

def solve_1(s: str) -> str:
    freq = [0] * 26
    for c in s:
        freq[ord(c) - ord('a')] += 1
    max_heap = [(-freq[i], chr(ord('a') + i)) for i in range(26) if freq[i] > 0]
    heapq.heapify(max_heap)
    res = []
    while len(max_heap) >= 2:
        f1, c1 = heapq.heappop(max_heap)
        f2, c2 = heapq.heappop(max_heap)
        res.append(c1)
        res.append(c2)
        if f1 + 1 < 0:
            heapq.heappush(max_heap, (f1 + 1, c1))
        if f2 + 1 < 0:
            heapq.heappush(max_heap, (f2 + 1, c2))
    if max_heap:
        f, c = max_heap[0]
        if -f > 1:
            return ""
        res.append(c)
    return "".join(res)

# ─────────────────────────────────────────────
# APPROACH 2: Interleave Even/Odd Indices | O(n log n) time | O(n) space
# EXPLAIN: Sort chars by frequency descending, fill even indices first then odd.
# WHEN: When you want a simple index-based placement without a heap.

def solve_2(s: str) -> str:
    freq = [0] * 26
    for c in s:
        freq[ord(c) - ord('a')] += 1
    max_freq = max(freq)
    if max_freq > (len(s) + 1) // 2:
        return ""
    pairs = sorted(
        [(freq[i], chr(ord('a') + i)) for i in range(26) if freq[i] > 0],
        reverse=True
    )
    res = [''] * len(s)
    idx = 0
    for f, c in pairs:
        for _ in range(f):
            if idx >= len(s):
                idx = 1
            res[idx] = c
            idx += 2
    return "".join(res)

# ─────────────────────────────────────────────
# APPROACH 3: Greedy with Previous Tracking | O(n log 26) time | O(1) space
# EXPLAIN: Pop the max-frequency char; if it equals the last placed char, use the second-highest instead.
# WHEN: Mirrors real-scheduler logic; more intuitive for step-by-step tracing.

def solve_3(s: str) -> str:
    freq = [0] * 26
    for c in s:
        freq[ord(c) - ord('a')] += 1
    max_freq = max(freq)
    if max_freq > (len(s) + 1) // 2:
        return ""
    max_heap = [(-freq[i], chr(ord('a') + i)) for i in range(26) if freq[i] > 0]
    heapq.heapify(max_heap)
    res = []
    prev_f, prev_c = 0, ''
    while max_heap:
        f, c = heapq.heappop(max_heap)
        if c == prev_c and max_heap:
            f2, c2 = heapq.heappop(max_heap)
            res.append(c2)
            if f2 + 1 < 0:
                heapq.heappush(max_heap, (f2 + 1, c2))
            heapq.heappush(max_heap, (f, c))
            prev_f, prev_c = f2 + 1, c2
        else:
            res.append(c)
            if f + 1 < 0:
                heapq.heappush(max_heap, (f + 1, c))
            prev_f, prev_c = f + 1, c
    return "".join(res) if len(res) == len(s) else ""

# ─────────────────────────────────────────────
# APPROACH 4: Counting + Direct Placement | O(n) time | O(n) space
# EXPLAIN: Place the most frequent character at all even positions first, then fill remaining positions.
# WHEN: Fastest O(n) approach; no sorting needed, just a single linear scan.

def solve_4(s: str) -> str:
    freq = [0] * 26
    for c in s:
        freq[ord(c) - ord('a')] += 1
    max_char = freq.index(max(freq))
    if freq[max_char] > (len(s) + 1) // 2:
        return ""
    res = [''] * len(s)
    idx = 0
    # Place most frequent at even slots
    while freq[max_char] > 0:
        res[idx] = chr(ord('a') + max_char)
        idx += 2
        freq[max_char] -= 1
    # Fill remaining chars
    for i in range(26):
        while freq[i] > 0:
            if idx >= len(s):
                idx = 1
            res[idx] = chr(ord('a') + i)
            idx += 2
            freq[i] -= 1
    return "".join(res)

# Made with Bob

File: public/java/089-reorganize-string.java
/*
 * LeetCode Problem #767: Reorganize String
 * Link: https://leetcode.com/problems/reorganize-string/
 * Difficulty: Medium
 */
import java.util.*;

class Solution {

    // APPROACH 1: Max Heap Greedy (Pair Pop) | O(n log 26) time | O(1) space
    // EXPLAIN: Each step pop the two most frequent chars, append both, re-push with decremented counts.
    public String reorganizeString1(String s) {
        int[] freq = new int[26];
        for (char c : s.toCharArray()) freq[c - 'a']++;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        for (int i = 0; i < 26; i++) if (freq[i] > 0) pq.offer(new int[]{freq[i], i});
        StringBuilder res = new StringBuilder();
        while (pq.size() >= 2) {
            int[] a = pq.poll(), b = pq.poll();
            res.append((char)('a' + a[1]));
            res.append((char)('a' + b[1]));
            if (a[0] - 1 > 0) pq.offer(new int[]{a[0] - 1, a[1]});
            if (b[0] - 1 > 0) pq.offer(new int[]{b[0] - 1, b[1]});
        }
        if (!pq.isEmpty()) {
            int[] last = pq.poll();
            if (last[0] > 1) return "";
            res.append((char)('a' + last[1]));
        }
        return res.toString();
    }

    // APPROACH 2: Interleave Even/Odd Indices | O(n log n) time | O(n) space
    // EXPLAIN: Sort chars by frequency descending; fill even indices first, then odd indices.
    public String reorganizeString2(String s) {
        int n = s.length();
        int[] freq = new int[26];
        for (char c : s.toCharArray()) freq[c - 'a']++;
        int maxFreq = 0;
        for (int f : freq) maxFreq = Math.max(maxFreq, f);
        if (maxFreq > (n + 1) / 2) return "";
        Integer[] indices = new Integer[26];
        for (int i = 0; i < 26; i++) indices[i] = i;
        Arrays.sort(indices, (a, b) -> freq[b] - freq[a]);
        char[] res = new char[n];
        int idx = 0;
        for (int i : indices) {
            for (int j = 0; j < freq[i]; j++) {
                if (idx >= n) idx = 1;
                res[idx] = (char)('a' + i);
                idx += 2;
            }
        }
        return new String(res);
    }

    // APPROACH 3: Greedy with Previous Tracking | O(n log 26) time | O(1) space
    // EXPLAIN: Always place the highest-frequency char; if it matches the last placed, use the second-highest.
    public String reorganizeString3(String s) {
        int[] freq = new int[26];
        for (char c : s.toCharArray()) freq[c - 'a']++;
        int maxFreq = 0;
        for (int f : freq) maxFreq = Math.max(maxFreq, f);
        if (maxFreq > (s.length() + 1) / 2) return "";
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        for (int i = 0; i < 26; i++) if (freq[i] > 0) pq.offer(new int[]{freq[i], i});
        StringBuilder res = new StringBuilder();
        int prevChar = -1;
        while (!pq.isEmpty()) {
            int[] top = pq.poll();
            if (top[1] == prevChar && !pq.isEmpty()) {
                int[] second = pq.poll();
                res.append((char)('a' + second[1]));
                if (second[0] - 1 > 0) pq.offer(new int[]{second[0] - 1, second[1]});
                pq.offer(top);
                prevChar = second[1];
            } else {
                res.append((char)('a' + top[1]));
                if (top[0] - 1 > 0) pq.offer(new int[]{top[0] - 1, top[1]});
                prevChar = top[1];
            }
        }
        return res.length() == s.length() ? res.toString() : "";
    }

    // APPROACH 4: Counting + Direct Placement | O(n) time | O(n) space
    // EXPLAIN: Place the most frequent char at all even slots first, then fill remaining slots sequentially.
    public String reorganizeString4(String s) {
        int n = s.length();
        int[] freq = new int[26];
        for (char c : s.toCharArray()) freq[c - 'a']++;
        int maxChar = 0;
        for (int i = 1; i < 26; i++) if (freq[i] > freq[maxChar]) maxChar = i;
        if (freq[maxChar] > (n + 1) / 2) return "";
        char[] res = new char[n];
        int idx = 0;
        while (freq[maxChar] > 0) { res[idx] = (char)('a' + maxChar); idx += 2; freq[maxChar]--; }
        for (int i = 0; i < 26; i++) {
            while (freq[i] > 0) {
                if (idx >= n) idx = 1;
                res[idx] = (char)('a' + i); idx += 2; freq[i]--;
            }
        }
        return new String(res);
    }
}

// Made with Bob
